### PR TITLE
[DROP-IN] Refactor: State centralization using Store and Reducers (FLUX architecture)

### PR DESCRIPTION
**Pull request by tomaszrybakiewicz**
Closes [#1596](https://github.com/mapbox/navigation-sdks/issues/1596)

### Problem
Current implementation splits DropIn UI state between multiple UIViewModel classes. 

This fragmentation of state forces us to: 
- inject multiple view models to a single UIComponent
- make each UIComponent aware of each UIViewModel responsibilities (know which UIViewModel holds which _state_ and processes which _action_)
- create UIViewModel <-> UIViewModel relationships (eg. `TripSessionStarterViewModel`, `AudioGuidanceViewModel`)

drop in - viewmodel current

drop in redux before

### Solution proposed in this PR 

Implement [FLUX architecture](https://facebook.github.io/flux/docs/in-depth-overview) by:
- implementing single Store class that centralizes Drop-In UI _state_ and _action_ dispatching
- updating each UIComponent to depend only on the Store
- decoupling ViewModels from one another and refactoring them act as Store Reducers

drop in - viewmodel new

drop in redux after

### QA App screen recording

https://user-images.githubusercontent.com/2678039/160676123-9ef4145c-fb60-4f1c-a3e4-9243c1f9bd77.mp4

### Tip: How to review this PR in Android Studio?

<details><summary>GIT: Stash current work and simulate PR merge by using --squash option</summary>

```bash
$ git stash
$ git checkout main
$ git pull origin main --ff-only
$ git pull origin tr-drop-in-store --squash
```
</details>

<details><summary>Android Studio: Open the project and switch to "All Changed Files" view</summary>
Screen Shot 2022-03-31 at 11 47 53 AM

</details>

<details><summary>GIT: Cleanup and restore previous work</summary>

```bash
$ git reset .
$ git checkout .
$ git clean -fd
$ git stash pop
```

</details>

### CODE REVIEW

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationView.kt
Comment (kmadsen) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationView.kt:70]: What is the reason this is here? seems like it should be with all the other state

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/lifecycle/UIViewModel.kt
Comment (kmadsen) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/lifecycle/UIViewModel.kt:35]: Why not delete the `UIViewModel`?

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/camera/CameraViewModel.kt
Comment (kmadsen) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/camera/CameraViewModel.kt:29]: Wondering if we should move this routing to the `Store`

Then the store would have a big `when`

``` kotlin
        return when (action) {
            is CameraAction -> cameraReducer.process(state.camera, action)
...
```

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/binder/ActionButtonBinder.kt
Comment (kmadsen) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/binder/ActionButtonBinder.kt:46]: I like how much simpler this is! We had this originally, but separated them with the hope that they would be more portable. For example, someone could use the LocationComponent without needing all the state

Here is another approach that I've been considering. That way it is a "module" that defines which dependency is provided to the component.

Rather than the component being responsible for deciding what it needs from the context

``` kotlin
return navigationListOf(
  module.audioGuidanceComponent(binding.soundButton),
  module.cameraModeButtonComponent(binding.cameraModeButton),
  module.recenterButtonComponent(binding.recenterButton),
)
```

``` kotlin
UIModule {
  fun recenterButtonComponent = RecenterButtonComponent(
     viewModel.cameraViewModel,
     viewModel.navigationStateViewModel,
     recenterButton
  )
...
```

Comment (tomaszrybakiewicz) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/binder/ActionButtonBinder.kt:46]: I was thinking the same but named ComponentFactory.

```kotlin
internal class ComponentFactory(
    private val context: DropInNavigationViewContext
) {
    fun locationComponent(mapView: MapView) = LocationComponent(context, mapView)

    fun routeLineComponent(mapView: MapView, lineOptions: MapboxRouteLineOptions) =
        RouteLineComponent(context, mapView, lineOptions)

    fun cameraComponent(mapView: MapView) = CameraComponent(context, mapView)

    fun mapMarkersComponent(mapView: MapView) = MapMarkersComponent(context, mapView)

    fun routeArrowComponent(mapView: MapView, arrowOptions: RouteArrowOptions) =
        RouteArrowComponent(mapView, arrowOptions)
}
```
usage:

```kotlin
internal class DropInNavigationViewContext(...) {
 ...
 fun componentFactory() = ComponentFactory(this)
}
```
```kotlin
internal class ActiveGuidanceMapBinder(
    private val context: DropInNavigationViewContext,
) : Binder<MapView> {

    override fun bind(mapView: MapView): MapboxNavigationObserver {
        context.componentFactory().apply {
            return navigationListOf(
                locationComponent(mapView),
                reloadOnChange(
                    context.mapStyleLoader.loadedMapStyle,
                    context.options.routeLineOptions
                ) { _, lineOptions ->
                    routeLineComponent(mapView, lineOptions)
                },
                cameraComponent(mapView),
                mapMarkersComponent(mapView),
                reloadOnChange(
                    context.mapStyleLoader.loadedMapStyle,
                    context.options.routeArrowOptions
                ) { _, arrowOptions ->
                    routeArrowComponent(mapView, arrowOptions)
                }
            )
        }
    }
}
```

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/camera/CameraViewModel.kt
Comment (tomaszrybakiewicz) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/camera/CameraViewModel.kt:29]: The Store must only know about  the `Action` and the `Reducer` interfaces.

Moving CameraAction -> cameraReducer routing to the Store will tightly couple Store to those concrete classes.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationView.kt
Comment (tomaszrybakiewicz) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationView.kt:74]: Yes. 
Adding `isReplayEnabled ` property was necessary for `MapboxNavigationViewCustomizedActivity` to be able to read and update `TripSessionStarterState.isReplayEnabled` that is buried deep inside the Store which is internal.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/binder/ActionButtonBinder.kt
Comment (kmadsen) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/binder/ActionButtonBinder.kt:46]: 👍  a Factory is good. 

Other point here if we require the `LocationComponent` to have an instance of `DropInNavigationViewContext`. Then we will not be able to create `LocationComponent` outside of the drop in ui.

But with the preference to take full adoption of `DropInNavigationViewContext`. Am thinking it makes sense to rename `DropInNavigationViewContext` to something more permanent. `DropInContext` for example.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationView.kt
Comment (kmadsen) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationView.kt:74]: https://github.com/mapbox/mapbox-navigation-android/pull/5626/files#diff-1181d521b74e76ffaacfed0d26a9c59920179338a76b4c61d1b0a7ad46437ca3R145

There is a TODO above that `MapboxNavigationViewCustomizedActivity` to create a better interface for surfacing state in `DropInNavigationView`. The current solution is dirty, but I don't think this change is solving it

We should not surface the `Store` state like this inside DropInNavigationView.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/model/Store.kt
Comment (kmadsen) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/model/Store.kt:56]: Adding a concern that there may be too much noise here.

For example, the `State.location` happens every second. In some cars it happens 10 times a second. Decoupling some states may be necessary at some point.

But we are notifying all reducers of state changes that only 1 of them care about. If we scale to have 50 reducers, and locations happening 10 times a second.. Do you have concerns about this loop becoming wasteful?

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationViewModel.kt
Comment (kmadsen) [inline libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationViewModel.kt:27]: I am a fan of centralizing state. But there are some issues have led us to decentralize it. Am raising one of the issues here.

All the `*ViewModel`s instances are "shareable" with android auto. This `Store` is only being instantiated with the ViewModel at this time, which we don't have in android auto.

The mechanism to share these viewmodels state with android auto is `MapboxNavigationObserver` through the `MapboxNavigationApp`. How would we share the `Store` state and the viewmodels with android auto with this? Why not make the `Store` state be a `MapboxNavigationObserver`?

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/model/Store.kt
Comment (tomaszrybakiewicz) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/model/Store.kt:56]: In FLUX every reducer receives every dispatched action. 

Of course we can optimize this, but I'd rather wait until we identify that this is an actual bottleneck. 

One potential improvement/optimization could be introduction of a convenience method that would allow reducer registration with predicate.

```kotlin 
fun register(predicate: (Action) -> Boolean, reducer: Reducer): Registration {
    val r = Reducer { state, action ->
        if (predicate(action)) reducer.process(state, action)
        else state
    }
    reducers.add(r)
    return Registration {
        reducers.remove(r)
    }
}

fun interface Registration {
    fun unregister()
}
```

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationViewModel.kt
Comment (tomaszrybakiewicz) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationViewModel.kt:27]: Sharing a state with android auto is a different problem that needs to be addressed in a separate PR.
This PR addresses state fragmentation issue.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/binder/ActionButtonBinder.kt
Comment (tomaszrybakiewicz) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/binder/ActionButtonBinder.kt:46]: Or, we can make `ComponentFactory` public and introduce a method somewhere that will return it pre-wired for external use.

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationViewModel.kt
Comment (kmadsen) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationViewModel.kt:27]: What's the proposed solution to share state with android auto after this change?

I had a clear idea of what to do when using instances of `MapboxNavigationObserver`, but after this change I think we'd need to change `Store` to `MapboxNavigationObserver` 🤔

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/binder/ActionButtonBinder.kt
Comment (kmadsen) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/binder/ActionButtonBinder.kt:46]: Yeah I agree that it would be good to make the ComponentFactory public. Unclear to me how we would expose the `ComponentFactory` without an instance of `DropInNavigationView` if it depends on `DropInNavigationViewContext` 🤔 

Right now you can create a `LocationComponent` but you first need to create a `LocationViewModel`. You do not need to create the entire `DropInNavigationView`

File: libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationViewModel.kt
Comment (tomaszrybakiewicz) [reply libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/DropInNavigationViewModel.kt:27]: I didn't look into it yet so it's hard for me to propose the solution for this.

Few ideas that come to mind:
- make Store singleton
- move Store to a service that's accessible by both Android Auto and Navigation View
- given MapboxNavigationObserver survives activity configuration changes... wrap store in some sort of StoreHolder class that can be registered with MapboxNavigationApp

**Comment by abhishek1508** [comment]
This is a great initiative and there are good improvements brought forward with this PR in terms of a central state management. Based on a small team discussion we need to have some plan of actions for the following items:

- What changes need to happen to make the `Component` (Ex: CameraComponent) to live independently outside the scope of Drop-In UI
- What changes need to happen to make the Store be shareable across Android-Auto
- Should we make the store instance publicly accessible so that the user can observe state changes and dispatch actions
- How do we feel about exposing some sort of `ComponentFactory` interface that would abstract the logic behind creating a component

Answers to some of these questions will become more clear once we have some prototype implementation of Android-Auto using these components from Drop-In UI as they are today. Let's not rush and try to find the answers to these questions to discuss some solutions that could be taken as a follow up immediately after this PR is merged to main

**Comment by abhishek1508** [comment]
> https://github.com/mapbox/mapbox-navigation-android/pull/5626#issuecomment-1085134180

Based on our discussion early last week, we all agreed to the following:
- DO NOT MERGE this PR before MVP release. 
- Merge this PR after MVP. 
- Following the merge, we need to address some issues and make this system so that we could host the UIComponents outside `libnavui-dropin` for Android-Auto to reuse them. As this is achieved, these components could be automatically used by the customers as well.
- It's ok if Android-Auto and customers would have to duplicate some logic. This seems the most logical way to keep the components segregated and not associate them with Drop-In UI specific architecture.

Next actions are captured [here](https://github.com/mapbox/navigation-sdks/issues/1639)

I will hold off from merging this today, so please feel free to raise concerns if any. This branch will be merged first thing tomorrow.

/cc @LukasPaczos @kmadsen @tomaszrybakiewicz

**Review by abhishek1508 — APPROVED** [review]
The changes look good to me. Let's merge and iterate.
The CI is failing @tomaszrybakiewicz

**Comment by tomaszrybakiewicz** [comment]
@abhishek1508 I've rebased on the latest `main`, fixed all merge conflicts and failing tests, and updated metalava API file.

**Pull request merged**
